### PR TITLE
Expose cabal-install lib, remove dogfooding, expose cabal-install-solver as own lib, enable HPC coverage reporting. 

### PR DESCRIPTION
**Pull request by emilypi**
As discussed in https://github.com/haskell/cabal/pull/7344#issuecomment-819670133, this PR does the following: 

- Exposes `cabal-install` as a library, bumping the `cabal-install` cabal version to 2.2, and removing the dogfooding framework surrounding `cabal-install` to facilitate easier testing. 
- Moves `cabal-install-solver` out to its own directory, and treats it as its own separate project. Addresses #6090, #7224, and #3781
- Enable HPC coverage: absorb `cabal-install-solver-dsl` module into the test modules, and incorporate it into its own test target. Splits out unit tests from long-running tests. Also adds proper subcomponent coverage masking so we can finally generate baseline coverage reports for `cabal-install`. 

TODO:
 
- [x] Doc changes for testing.md, contributing.md
- [x] Changelog entry 
- [x] Changes to CI 
 
Fixes: 

- #7224 
- #6090
- #4798 
- #3781
- #1597

Baseline report: 

[cabal-install-baseline-coverage.zip](https://github.com/haskell/cabal/files/6313866/cabal-install-baseline-coverage.zip)
 

CC @fgaz @phadej @ptkato

### CODE REVIEW

File: cabal-install/cabal-install.cabal [hunk 1]
```diff
@@ -12,68 +9,46 @@ Description:
...
-      if impl(ghc < 8.8)
-        ghc-options: -Wnoncanonical-monadfail-instances
 
+library
+    default-language: Haskell2010
+    ghc-options: -Wall -Wcompat -Wnoncanonical-monad-instances
+    if impl(ghc < 8.8)
+      ghc-options: -Wnoncanonical-monadfail-instances
       if impl(ghc >=8.10)
         ghc-options: -Wunused-packages
```
Comment (fgaz) [inline cabal-install/cabal-install.cabal:48]: What about using common stanzas now that we can? So that we only need to update this stuff in one place

File: cabal-install-solver/cabal-install-solver.cabal [hunk 2]
```diff
@@ -39,18 +39,18 @@ library
   default-language: Haskell2010
   hs-source-dirs:   src
   hs-source-dirs:   src-assertion
-  ghc-options:      -Wall -fwarn-tabs -fwarn-incomplete-uni-patterns
-
-  if impl(ghc >=8.0)
```
Comment (fgaz) [inline cabal-install-solver/cabal-install-solver.cabal:44]: I think this and the other ghc compat stuff in this commit should be moved to the previous one, where the minimum ghc actually changed (or to a new commit)

Comment (fgaz) [reply cabal-install-solver/cabal-install-solver.cabal:44]: Oh, I see that file was `cabal-version: 2.2` even before the change in cabal-install. So I guess a new commit?

File: cabal-install/cabal-install.cabal (hunk 1 above)
Comment (emilypi) [reply cabal-install/cabal-install.cabal:48]: Oo! good point. In the previous iteration i had `cabal-version: 2.0`, but `2.2` seemed like a better idea for internal libs. We definitely can do common stanzas.

File: cabal-install-solver/cabal-install-solver.cabal (hunk 2 above)
Comment (emilypi) [reply cabal-install-solver/cabal-install-solver.cabal:44]: Yes, it was changed upon exposing the new library to match the `cabal-install.dev` cabal file.

File: cabal-install/cabal-install.cabal (hunk 1 above)
Comment (emilypi) [reply cabal-install/cabal-install.cabal]: Done

**Review by fgaz — COMMENTED** [review]
The workflow files are generated (see the `github-actions` make target)

edit: this was supposed to be a comment on commit 59b2fe57188b6de6eb62ee993be083d40db4a7ec

**Comment by fgaz** [comment]
Could you squash the commits that do the same thing (eg. Setup.hs for cabal-install-solver → separate cabal-install-solver)?

**Comment by emilypi** [comment]
Sure i'll restructure these and rebase tomorrow 👍

File: cabal-install-solver/cabal-install-solver.cabal [hunk 3]
```diff
@@ -39,18 +39,18 @@ library
...
-    if impl(ghc <8.8)
-      ghc-options: -Wnoncanonical-monadfail-instances
+  ghc-options:
+    -Wall -Wcompat -Wnoncanonical-monad-instances
+    -fwarn-tabs -fwarn-incomplete-uni-patterns
 
+  if impl(ghc <8.8)
+    ghc-options: -Wnoncanonical-monadfail-instances
     if impl(ghc >=8.10)
       ghc-options: -Wunused-packages
```
Comment (fgaz) [inline cabal-install-solver/cabal-install-solver.cabal:49]: The inner condition is always false. I think it's one level too deep

File: cabal-install/cabal-install.cabal [hunk 4]
```diff
@@ -12,68 +9,51 @@ Description:
...
-                     -Wnoncanonical-monad-instances
-      if impl(ghc < 8.8)
-        ghc-options: -Wnoncanonical-monadfail-instances
-
+common warnings
+    ghc-options: -Wall -Wcompat -Wnoncanonical-monad-instances
+    if impl(ghc < 8.8)
+      ghc-options: -Wnoncanonical-monadfail-instances
       if impl(ghc >=8.10)
         ghc-options: -Wunused-packages
```
Comment (fgaz) [inline cabal-install/cabal-install.cabal:46]: Same here

File: cabal-install/cabal-install.cabal [hunk 5]
```diff
@@ -205,78 +185,21 @@ executable cabal
...
-
     build-depends:
         async      >= 2.0      && < 2.3,
         array      >= 0.4      && < 0.6,
-        base       >= 4.8      && < 4.15,
         base16-bytestring >= 0.1.1 && < 1.1.0.0,
         binary     >= 0.7.3    && < 0.9,
         bytestring >= 0.10.6.0 && < 0.12,
         Cabal      == 3.5.*,
+        cabal-install-solver,
```
Comment (fgaz) [inline cabal-install/cabal-install.cabal:195]: I think this needs a constraint like the Cabal one above (assuming we release cabal-install-solver in lockstep with Cabal and cabal-install)

File: cabal-install/src/Distribution/Client/Utils.hs [hunk 6]
```diff
@@ -362,8 +360,4 @@ progressMessage verbosity phase subject = do
         ProgressCompleted   -> "Completed    "
 
 cabalInstallVersion :: Version
-#ifdef __DOCTEST__
-cabalInstallVersion = mkVersion [3,3]
-#else
-cabalInstallVersion = mkVersion' Paths_cabal_install.version
-#endif
+cabalInstallVersion = mkVersion [3,5]
```
Comment (fgaz) [inline cabal-install/src/Distribution/Client/Utils.hs:363]: Hmm we probably need to add this to the release checklist (can be in another pr)

File: cabal-install/cabal-install.cabal [hunk 7]
```diff
@@ -205,78 +185,21 @@ executable cabal
...
         containers >= 0.5.6.2  && < 0.7,
         cryptohash-sha256 >= 0.11 && < 0.12,
         deepseq    >= 1.4.1.1  && < 1.5,
         directory  >= 1.2.2.0  && < 1.4,
         echo       >= 0.1.3    && < 0.2,
         edit-distance >= 0.2.2 && < 0.3,
+        exceptions,
```
Comment (fgaz) [inline cabal-install/cabal-install.cabal:202]: where is this needed? I don't see any change that uses it. It also needs bounds

File: cabal-install/cabal-install.cabal [hunk 8]
```diff
@@ -317,13 +236,171 @@ executable cabal
...
+          network-uri >= 2.6.2.0 && <2.6.4.0,
+          random,
+          tar,
+          time,
+          zlib,
+          tasty >= 1.2.3 && <1.4,
+          tasty-golden >=2.3.1.1 && <2.4,
+          tasty-quickcheck,
+          tasty-hunit >= 0.10,
+          tree-diff,
```
Comment (fgaz) [inline cabal-install/cabal-install.cabal:315]: tasty-quickcheck and tree-diff may also need a bound

File: cabal.project [hunk 9]
```diff
@@ -1,5 +1,5 @@
 packages: Cabal/ cabal-testsuite/
-packages: cabal-install/cabal-install-solver/
+packages: cabal-install-solver/
```
Comment (fgaz) [inline cabal.project:2]: This should be in the previous commit

File: cabal.project [hunk 10]
```diff
@@ -29,18 +29,56 @@ allow-newer: windns-0.1.0.1:base
...
+--
+-- NOTE: for library coverage in multi-project builds,
+-- see:
+--
+-- * https://github.com/haskell/cabal/issues/6440
+-- * https://github.com/haskell/cabal/issues/5213#issuecomment-586517129
+--
+-- We must mask coverage for dependencies of `cabal-install` in
+-- multiproject settings in order to generate coverage for
+-- the `cabal-install` library
```
Comment (fgaz) [inline cabal.project:49]: Question: how do we get the coverage for Cabal?

File: templates/ci-artifacts.template.yml [hunk 11]
```diff
@@ -27,8 +27,8 @@ jobs:
       - uses: actions/checkout@v2
       - name: Package project
         run: |
-          cabal v2-sdist --builddir=sdist Cabal cabal-install
-          python3.8 release.py --tarlib=sdist/sdist/Cabal-*.tar.gz --tarexe=sdist/sdist/cabal-install-*.tar.gz
+          cabal v2-sdist --builddir=sdist Cabal cabal-install-solver cabal-install
+          python3.8 release.py --tarlib=sdist/sdist/Cabal-*.tar.gz --tarsolver=sdist/sdist/cabal-install-solver-3.*.tar.gz --tarexe=sdist/sdist/cabal-install-3.*.tar.gz
```
Comment (fgaz) [inline templates/ci-artifacts.template.yml:31]: why the "3."s?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -26,54 +26,23 @@ cabal v2-build Cabal:unit-tests # build Cabal's unit test suite
 cabal v2-build cabal-tests # etc...
 ~~~~
 
-**Dogfooding HEAD.**
```
Comment (fgaz) [inline CONTRIBUTING.md:29]: This is still valid though, isn't it?

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -137,22 +105,16 @@ There are also other test suites:
...
+  cabal-install's dependency solver, VCS, and file monitoring code.
+  If you are working on the solver you should run this test suite.
 
 * `cabal-install:integration-tests2` are integration tests on some
   top-level API functions inside the `cabal-install` source code.
 
 For these test executables, `-p` which applies a regex filter to the test
 names.
 
-**Testing `cabal-install` Locally**
```
Comment (fgaz) [inline CONTRIBUTING.md:150]: I think this section should stay and just say "run cabal test". After all running tests locally is still encouraged!

File: changelog.d/pr-7358 [hunk 14]
```diff
@@ -0,0 +1,21 @@
+synopsis: Enable HPC coverage for `cabal-install`
```
Comment (fgaz) [inline changelog.d/pr-7358:1]: I think the main point _for users_ is the cabal-install-solver/cabal-install-library stuff more than hpc

**Review by fgaz — CHANGES_REQUESTED** [review]
Small stuff and questions

File: cabal-install/cabal-install.cabal (hunk 8 above)
Comment (fgaz) [reply cabal-install/cabal-install.cabal:315]: Nvm, this isn't new

File: cabal.project (hunk 10 above)
Comment (emilypi) [reply cabal.project:49]: Through pain and suffering. We'd have to rework the `cabal-testsuite`, and associated tests to be coverable. I'm punting on that one until i understand it better.

File: templates/ci-artifacts.template.yml (hunk 11 above)
Comment (emilypi) [reply templates/ci-artifacts.template.yml:31]: because now that `cabal-install-solver` exists, `cabal-install-*` matches two possible conditions. This allows us to match  on file names more precisely.

File: cabal-install/cabal-install.cabal (hunk 5 above)
Comment (emilypi) [reply cabal-install/cabal-install.cabal:195]: Good idea.

File: cabal-install/cabal-install.cabal (hunk 7 above)
Comment (emilypi) [reply cabal-install/cabal-install.cabal:202]: Artifact from the Cabal Init work. Nice catch - i'll remove.

File: CONTRIBUTING.md (hunk 12 above)
Comment (emilypi) [reply CONTRIBUTING.md:29]: I think it's confusing to keep, since it implies having to think about something special. It is now the _only_ way we test `cabal-install`, so i don't think it makes sense to point out specifically.

File: CONTRIBUTING.md (hunk 13 above)
Comment (emilypi) [reply CONTRIBUTING.md:150]: Same as above - calling `cabal test` is normal behavior, so i don't think it needs to be highlighted. We can debate this - I think it's unnecessary to point out details that people should _expect_ to already be a thing.

**Comment by typedrat** [comment]
This would be massively useful to me.

File: cabal-install-solver/cabal-install-solver.cabal (hunk 3 above)
Comment (emilypi) [reply cabal-install-solver/cabal-install-solver.cabal:49]: Fixed

File: cabal-install/src/Distribution/Client/Utils.hs (hunk 6 above)
Comment (emilypi) [reply cabal-install/src/Distribution/Client/Utils.hs:363]: Added

File: cabal.project (hunk 9 above)
Comment (emilypi) [reply cabal.project:2]: :+1:

**Comment by emilypi** [comment]
@fgaz okay - comments should be addressed, and I'll merge on green, following up with a PR with updates to the release checklist. The release checklist in particular needs to detail how to update CI, which i'll add as well. I'll do one more pass to see if anything stands out, and then I think we're good to merge.

**Review by fgaz — APPROVED** [review]

**Comment by emilypi** [comment]
Was this closed by linking? I didn't physically close this issue.
